### DISCUSSION TITLE
5.0.0 Migration Guide

### DISCUSSION THREAD

**Discussion by Keboo** [MaterialDesignInXAML/MaterialDesignInXamlToolkit#3466] [Announcements]
This is a living document that is expected to be a guide for updating from 4.9.0 to 5.0.0. All feedback is welcome here, however if you are running into a specific issue with the migration, please open a new issue rather than commenting here. 

This will cover the most common upgrade issues. You can find a list of breaking changes on issue: #2255. 
Anything that was previously marked as Obsolete in the 4.9.0 or earlier release has been removed.

# Main Defaults Resource Dictionary Renamed
Because we are still supporting both Material Design 2 and Material Design 3 styles, the previous `MaterialDesignTheme.Defaults.xaml` resource dictionary has been replaced with `MaterialDesign2.Defaults.xaml` and `MaterialDesign3.Defaults.xaml`. In your main App.xaml replace with one of those two dictionaries. If you want your app to still use the same previous look and feel, the `MaterialDesign2.Defaults.xaml` resource dictionary most closely resembles the defaults from the 4.9.0 release.

# Brush Renames
To better facilitate the addition of new brushes as well as to make adjusting individual components, the library has adopted a new naming scheme for brush resources. 

1. All brush resources will now start with the `MaterialDesign.Brush.` prefix.
2. Theme color brushes (Primary and Secondary) will be in the form `MaterialDesign.Brush.<Primary|Secondary>.[Light|Dark]`. Corresponding foreground colors, will have a `.Foreground` suffix added. For example `MaterialDesign.Brush.Secondary.Light` or `MaterialDesign.Brush.Primary.Foreground`.
3. "Accent" has been replaced with "Secondary" on all brushes.

For anyone wishing to avoid needing to do the brush rename at the time of upgrade, there is an additional resource dictionary that you can include that maps the old brush resource names to the new names. This would typically get placed in the merged resource dictionary of your App.xaml.
```xaml
<ResourceDictionary Source="pack://application:,,,/MaterialDesignThemes.Wpf;component/Themes/MaterialDesignTheme.ObsoleteBrushes.xaml" />
```

Be aware that this should be considered a _**temporary**_ solution to allow time to do the full migration. It is not expected to be maintained long term and is subject to have items removed from it on future releases.

There is also a migration script that can be run on your repository that will attempt to locate brush resources and update them. Be aware this only looks for `{StaticResource xxx}` and `{DynamicResource xxx}` references for brushes. You can find the script in the repository [here](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/blob/d065de911acc6a68b8594d19ad4cffde9519ef14/build/MigrateBrushes.ps1)

See #2435 for more additional details.

# Style renames
To standardize on naming, all usages of "Accent" in the style keys were replaced with "Secondary". For many style keys, simply making this one change will appropriately update it. 

1. All obsolete styles have been removed.
2. "MaterialDesignNavigatilRailTabControl" fixed the typo and is now "MaterialDesignNavigationRailTabControl"
3. 

For anyone wishing to avoid needing to do the style rename at the time of upgrade, there is an additional resource dictionary that you can include that maps the old styles keys to the new names. This would typically get placed in the merged resource dictionary of your App.xaml.
```xaml
<ResourceDictionary Source="pack://application:,,,/MaterialDesignThemes.Wpf;component/Themes/MaterialDesignTheme.ObsoleteStyles.xaml" />
```

Be aware that this should be considered a _**temporary**_ solution to allow time for you to do the full migration. It is not expected to be maintained long term and is subject to have items removed from it on future releases.

There is also a migration script that can be run on your repository that will attempt to locate style resources and update them. Be aware this only looks for `{StaticResource xxx}` and `{DynamicResource xxx}` style references. You can find the script in the repository [here](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/blob/d065de911acc6a68b8594d19ad4cffde9519ef14/build/MigrateStyles.ps1)

# ShadowAssist replaced by ElevationAssist
The ShadowAssist was previously marked as Obsolete in the 4.9.0 release and has been completely removed int he 5.0.0 release.
To fix any usages simply update from `ShadowAssist` to `ElevationAssist` and `ShadowDepth` to `Elevation`. The old values map to the new ones as follows:
`ShadowDepth.Depth0` => `Elevation.Dp0`
`ShadowDepth.Depth1` => `Elevation.Dp2`
`ShadowDepth.Depth2` => `Elevation.Dp3` or `Elevation.Dp4`
`ShadowDepth.Depth3` => `Elevation.Dp7` or `Elevation.Dp8`
`ShadowDepth.Depth4` => `Elevation.Dp12` or `Elevation.Dp16`
`ShadowDepth.Depth5` => `Elevation.Dp16` or `Elevation.Dp24`

# Code renames

All extension method classes have had their "Mixin" suffix replaced with an "Extensions" suffix. In most cases, this will be a transparent change, unless you are invoking the extension method directly as a static method.

**Comment by longkimkao** [comment]
Hello,  how can i change the height of textbox outline ?

**Reply by Keboo** [reply]
I assume you are looking to change the thickness of the outline? If so, the property you want to change is `BorderThickness`

**Reply by longkimkao** [reply]
Thank for your reply.
I mean reduce the height of this textbox, change the height from this
![2024-04-25_10-54-52](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/assets/110613954/b9b15fc4-9f00-4249-896b-90825a5e1680)
to this
![2024-04-25_11-03-06](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/assets/110613954/0a8501d6-abac-4448-a45c-e15cf2d3c0f3)
i use poperty height, fontsize, padding. But i thing MaterialDesign will have property to do that.

**Reply by Keboo** [reply]
It sounds like you are likely already on the right path. The `Padding` property is the biggest contributor toward that default height. To get a look similar to your picture you are probably looking to do about `Padding="8"` or similar.

**Comment by maydoc** [comment]
Hi, regarding ShadowAssist.ShadowEdges, I'm getting an error that it doesn't exist, what name was it changed to?

**Reply by Keboo** [reply]
It was functionally removed in #3217 and was removed in the 5.0.0 because it was no longer be respected by any of the controls. If it is still in use in your project, you can simply wrap the control and clip it to achieve the same visual effect.

**Comment by mohsen-id** [comment]
hi, What is the best way to access these PrimaryXXX colors in version 5?

**Reply by Keboo** [reply]
The brushes were renamed, but the colors are still available and can be referenced the same as before. However, be aware that all of the "Accent" resource dictionaries did have their names changed to be "Secondary" instead. Beyond that, all of the color resource dictionaries should be unchanged. You can see all of them here: https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/tree/master/MaterialDesignColors.Wpf/Themes

**Comment by mohsen-id** [comment]
Thanks for reply.
The link goes to 404 - page not found!

I migrated to version 5 and in some parts of the code I used Static Resource PrimaryXXX colors that do not exist now. If I add `MaterialDesignColor.XXX.xaml` as a ResourceDictionary to App.xaml, it affects BundledTheme and changes the Foreground color for example.

What is the best way to access these PrimaryXXX colors in version 5?

my App.xaml:
`<ResourceDictionary.MergedDictionaries> <md:BundledTheme BaseTheme="Light" PrimaryColor="DeepOrange" SecondaryColor="Lime" /> <ResourceDictionary Source="pack://application:,,,/MaterialDesignThemes.Wpf;component/Themes/MaterialDesign3.Defaults.xaml" /> </ResourceDictionary.MergedDictionaries>`

**Reply by Keboo** [reply]
My apologies on the link we just recently redid the repo structure, which broke the link. Here is the updated one.

https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/tree/master/src/MaterialDesignColors.Wpf/Themes


As for accessing the colors, this works for me, and I am not seeing the issue with the theme problems you describe.

App.xaml
```xaml
<Application.Resources>
  <ResourceDictionary>
    <ResourceDictionary.MergedDictionaries>
      <materialDesign:BundledTheme BaseTheme="Inherit"
                                   ColorAdjustment="{materialDesign:ColorAdjustment}"
                                   PrimaryColor="DeepPurple"
                                   SecondaryColor="Lime" />

      <ResourceDictionary Source="pack://application:,,,/MaterialDesignColors;component/Themes/MaterialDesignColor.Amber.xaml" />

      <ResourceDictionary Source="pack://application:,,,/MaterialDesignThemes.Wpf;component/Themes/MaterialDesign2.Defaults.xaml" />
    </ResourceDictionary.MergedDictionaries>
...
```

Then in my application:
```xaml
<Button Style="{StaticResource MaterialDesignFloatingActionMiniLightButton}"
        ToolTip="MaterialDesignFloatingActionMiniLightButton">
  <Button.Background>
    <SolidColorBrush Color="{DynamicResource Primary400}" />
  </Button.Background>
  <materialDesign:PackIcon Width="{StaticResource IconSize}"
                            Height="{StaticResource IconSize}"
                            Kind="Alarm" />
</Button>
```

In the demo app, this then looks something like this (first button is now using the Amber color).
![image](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/assets/952248/6ed86e72-4ddb-4271-a9a9-53107f7bdb53)

**Comment by longkimkao** [comment]
Hi Keboo, i'm trying to override the defalut style but getting many error like this.
![2024-07-13_09-55-43](https://github.com/user-attachments/assets/8eca3142-cabf-42e5-b300-caea85a320d1)
I installed nuget packet ver 5.0, any solution to this problem ?. 
I want to download the  .cs file and modify it but don't know how to do :(
Any help would be greatly appreciated !

**Reply by Keboo** [reply]
The issue here is that the [BindableItemsHost is declared internal](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/blob/5438e9cd42db7c18cf82f46d27ff39eabc928658/src/MaterialDesignThemes.Wpf/TabAssist.cs#L19-L34). What you want to do is download that file, change its namespace, and leverage that within your local copy of the template.

**Comment by bezo97** [comment]
Hi, I'm updating from 4.9 to 5.1 and there's a converter called `InverseBoolToVisConverter` that went missing. Might've changed namespace or something because I can still see it in the codebase. It's not a big deal though, but it would be worth mentioning it in the guide.

**Comment by corvinsz** [comment]
Hey @Keboo,

do you by chance still have the scripts to automatically convert from the old Brushes/Styles?
The [links](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/blob/master/Scripts/MigrateBrushes.ps1) you posted above are not working anymore:
![image](https://github.com/user-attachments/assets/fbbe265e-6560-44dd-be8f-8e02ecfe7f03)


Edit: I noticed the MD3 demo is still using alot of old/obsolete Brushes, hence the question for the script.

**Reply by Keboo** [reply]
Yes sorry, the repo refactor broke some of those. Here is the correct link. https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/blob/d065de911acc6a68b8594d19ad4cffde9519ef14/build/MigrateBrushes.ps1

I also updated my links above.
